src/newrelic.coffee

```diff
@@ -126,39 +135,90 @@ plugin = (robot) ->
       gist {content: messages, enterpriseOnly: true, fileExtension: 'md'}, (err, resp, data) ->
         robot.messageRoom room, "#{longMessageIntro} View output at: #{data.html_url}"
 
+  subscriptions_for_subscriber = (subscriber) ->
+    subscription_lookup = robot.brain.get "newrelicviolations_subscriptions"
+    (sub for sub_id, sub of subscription_lookup when sub.subscriber == subscriber )
+
+  setting_for_channel = (channel, variable) ->
+    all_channel_settings  = robot.brain.get "newrelicalerts_channelsettings"
+    settings = all_channel_settings[channel] or {}
+    settings[variable]
+
+
+
+  annotate_incident_links = (violations, incident_lookup) ->
+    account_id = process.env.HUBOT_NEWRELIC_INSIGHTS_API_ENDPOINT.match(/accounts\/(\d+)/i)[1]
+    for violation in violations
+      incident_id = incident_lookup[violation.id]
+      if incident_id
+        violation['url'] = "https://alerts.newrelic.com/accounts/#{account_id}/incidents/#{incident_id}/violations"
+      else
+        # fail over to the policy page
+        violation['url'] = "https://alerts.newrelic.com/accounts/#{account_id}/policies/#{violation.links.policy_id}"
+
+    return violations
+
+  violations_for_subscriber = (all_violations, subscriber) ->
+    subs = subscriptions_for_subscriber subscriber
+    violations = []
+    for v in all_violations
+      for sub in subs
+        if v[sub.field] and minimatch v[sub.field], sub.pattern
+          violations.push v
+    return violations
+
+  dispatch_to_subscribers = (robot, alerts, action) ->
+      subscription_lookup = robot.brain.get "newrelicviolations_subscriptions"
+      updated_channels = []
+      for alert in alerts
+        destinations = _.uniq (sub_details.subscriber for sub_id, sub_details of\
+          subscription_lookup\
+          when alert[sub_details.field] and minimatch alert[sub_details.field], sub_details.pattern)
+        for channel in destinations
+          updated_channels.push channel
+          robot.messageRoom channel, "Alert #{action}: #{alert.policy_name} | #{alert.condition_name} ([#{alert.id}](#{alert.url}))"
+      return updated_channels
+
   poll_violations = (robot) ->
+    get "alerts_incidents.json", (err, json)->
```

Comment:
I'm not too clear on the order of operations here, but since the second `get` in this function uses some data that gets set in the brain in the first `get`, and since I assume the first `get` is non-blocking and could complete after the second `get`, I wonder if there would be any reason to chain them together (via promises or some other means) to ensure that the first one completes first. Again, not sure if this is desirable, but wanted to raise the possibility in case it is.

Reply:
I didn't realize that these calls were asynchronous-- it probably does call for a bit of re-architecting in that case.

Reply:
I fixed this by enclosing the violations callback in the incidents callback